gems/sorbet-runtime/tools/generate_call_validation.cc

```diff
@@ -0,0 +1,261 @@
+
+struct Options {
+    bool bindCall;
+};
+
+enum class ValidatorKind {
+    Method,
+    Procedure,
+};
+
+string_view validatorKindToString(ValidatorKind kind) {
+    switch (kind) {
+        case ValidatorKind::Method:
+            return "method"sv;
+        case ValidatorKind::Procedure:
+            return "procedure"sv;
+    }
+}
+
+constexpr size_t MAX_ARITY = 4;
+
+variant<Options, int> parseArgs(int argc, char **argv) {
+    cxxopts::Options options("generate_call_validation");
+    options.add_options()("bind-call", "Whether to use UnboundMethod#bind_call or not", cxxopts::value<bool>());
+    options.add_options()("h,help", "Show this help");
+
+    try {
+        auto raw = options.parse(argc, argv);
+        if (raw["help"].as<bool>()) {
+            fmt::print(stderr, "{}", options.help());
+            return 0;
+        }
+
+        if (raw["bind-call"].count() == 0) {
+            fmt::print(stderr, "Missing required option: --bind-call\n\n{}", options.help());
+            return 1;
+        }
+
+        return Options{
+            raw["bind-call"].as<bool>(),
+        };
+    } catch (cxxopts::OptionParseException &e) {
+        fmt::print(stderr, "{}\n\n{}", e.what(), options.help());
+        return 1;
+    }
+}
+
+void generateCreateValidatorFastDispatcher(ValidatorKind kind) {
+    auto kindString = validatorKindToString(kind);
+    fmt::print("  def self.create_validator_{}_fast(mod, original_method, method_sig)\n", kindString);
+    switch (kind) {
+        case ValidatorKind::Method:
+            fmt::print("    if method_sig.return_type.is_a?(T::Private::Types::Void)\n");
+            fmt::print("      raise 'Should have used create_validator_procedure_fast'\n");
+            fmt::print("    end\n");
+            break;
+        case ValidatorKind::Procedure:
+            break;
+    }
+
+    const char *returnTypeArg;
+    switch (kind) {
+        case ValidatorKind::Method:
+            returnTypeArg = ", method_sig.return_type.raw_type";
+            break;
+        case ValidatorKind::Procedure:
+            returnTypeArg = "";
+            break;
+    }
+    fmt::print("    # trampoline to reduce stack frame size\n");
+    for (size_t arity = 0; arity <= MAX_ARITY; arity++) {
+        if (arity == 0) {
+            fmt::print("    if method_sig.arg_types.empty?\n");
+        } else {
+            fmt::print("    elsif method_sig.arg_types.length == {}\n", arity);
+        }
+
+        fmt::print("      create_validator_{}_fast{}(mod, original_method, method_sig{}", kindString, arity,
+                   returnTypeArg);
+        for (size_t i = 0; i < arity; i++) {
+            fmt::print(",\n");
+            fmt::print("                                    method_sig.arg_types[{}][1].raw_type", i);
+        }
+        fmt::print(")\n");
+    }
+    fmt::print("    else\n");
+    fmt::print("      raise 'should not happen'\n");
+    fmt::print("    end\n");
+    fmt::print("  end\n");
+    fmt::print("\n");
```

Comment:
Slight preference for using concatenated string literals:

```c++
fmt::print("  else\n"
           "    raise 'should not happen'\n"
           "  end\n"
           ...);
```

or use the new-fangled raw-string literals: https://en.cppreference.com/w/cpp/language/string_literal

you can even do a (typically C++ ugly) here-doc sort of thing to avoid having to escape anything:

```c++
fmt::print(R"STRLIT(    else
        raise "should not happen"
      end
 ....)STRLIT")
```

I have a small preference for the former, but either one is easier to read than a long string of `fmt::print` calls.

Reply:
😅

I can make this change, but I think it's a worse structuring,

- variables end up way farther down, at the bottom of many lines, instead of on the line they're being interpolated into
- if you happen to not end a line with `\n` clang-format will have its way with formatting the code
- for the same reason why we prefer trailing commas, i prefer this format (e.g., the diff from inserting one line at the start or at the end is small, you can copy / paste either the start or the end line to add one more line and start changing it)

in particular i think the first point (location of variable interpolation) is a deal breaker for me, so i'll only coallesce `fmt::print` lines where there isn't an intervening interpolation.

Reply:
> in particular i think the first point (location of variable interpolation) is a deal breaker for me, so i'll only coallesce `fmt::print` lines where there isn't an intervening interpolation.

Your explanation makes sense to me and is what I had in mind, though I failed to communicate it adequately. :+1: